wgpu-hal/src/vulkan/mod.rs

```diff
@@ -359,18 +463,70 @@ pub struct Device {
     render_doc: crate::auxil::renderdoc::RenderDoc,
 }
 
+/// Semaphores that a given submission should wait on and signal.
+struct RelaySemaphoreState {
+    wait: Option<vk::Semaphore>,
+    signal: vk::Semaphore,
```

Comment:
My reading is that you can actually just use one semaphore to order submits, for both the signal and wait operations.

The rules I can find about signaling and waiting just say they have to happen in the natural order, and a submission will definitely wait before signalling. So the semaphore will always be in the right state for the next operation.

It's not a problem just to enqueue multiple operations that will wait and signal the same semaphore - if it were, then you couldn't handle an arbitrary number of submissions just by swapping two semaphores back and forth. Certainly when you're doing that, half the submissions are waiting on the same semaphore. So if that's permitted, you might as well just use one semaphore for everything.

The text of `VUID-VkSubmitInfo2-semaphore-03881` explicitly mentions the possibility of semaphores appearing in both lists on the same submit:

> If the same semaphore is used as the semaphore member of both an element of `pSignalSemaphoreInfos` and `pWaitSemaphoreInfos`, and that semaphore is a timeline semaphore, the value member of the `pSignalSemaphoreInfos` element must be greater than the value member of the `pWaitSemaphoreInfos` element

That's describing timeline semaphores, but it shows that the authors are expecting this possibility. And there's no VUID saying that the same semaphore can't appear in both lists.

Reply:
As discussed in Matrix (my bad), technically what I said above is correct, but we want to keep the semaphore-swapping code for the time being because of https://gitlab.freedesktop.org/mesa/mesa/-/issues/5508 .